src/overmap.cpp

```diff
@@ -1348,139 +1385,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
-            } else if (blink && npc_here) {
+            } else if (blink && overmap_buffer.has_npc(omx, omy, z)) {
                 // Display NPCs only when player can see the location
                 ter_color = c_pink;
-                ter_sym = '@';
-            } else if (blink && horde_here) {
+                ter_sym   = '@';
+            } else if (blink && los && overmap_buffer.has_horde(omx, omy, z)) {
                 // Display Hordes only when within player line-of-sight
                 ter_color = c_green;
-                ter_sym = 'Z';
-            } else if (blink && veh_here) {
+                ter_sym   = 'Z';
+            } else if (blink && overmap_buffer.has_vehicle(omx, omy, z)) {
                 // Display Vehicles only when player can see the location
                 ter_color = c_cyan;
-                ter_sym = 'c';
+                ter_sym   = 'c';
+            } else if (!sZoneName.empty() && tripointZone.x == omx && tripointZone.y == omy) {
+                ter_color = c_yellow;
+                ter_sym   = 'Z';
             } else {
-                if (sZoneName != "" && tripointZone.x == omx && tripointZone.y == omy) {
-                    ter_color = c_yellow;
-                    ter_sym = 'Z';
-                } else {
-                    // Nothing special, but is visible to the player.
-                    if (otermap.find(cur_ter) == otermap.end()) {
+                // Nothing special, but is visible to the player.
+                // First see if we have the oter_t cached
+                oter_t const* info = nullptr;
+                for (auto const &c : cache) {
+                    if (c.first == cur_ter) {
+                        info = c.second;
+                        break;
+                    }
+                }
+                // Nope, look in the hash map next
+                if (!info) {
+                    auto const it = otermap.find(cur_ter);
+                    if (it == otermap.end()) {
                         debugmsg("Bad ter %s (%d, %d)", cur_ter.c_str(), omx, omy);
                         ter_color = c_red;
-                        ter_sym = '?';
+                        ter_sym   = '?';
                     } else {
-                        // Map tile marked as explored
-                        if (show_explored && overmap_buffer.is_explored(omx, omy, z)) {
-                            ter_color = c_dkgray;
-                        } else {
-                            ter_color = otermap[cur_ter].color;
-                        }
-                        ter_sym = otermap[cur_ter].sym;
+                        // cache the new value
+                        info = &it->second;
+                        cache[cache_next] = std::make_pair(cur_ter, info);
+                        cache_next = (cache_next + 1) % cache_size;
                     }
                 }
+                // Ok, we found something
+                if (info) {
```

Comment:
minor critique, since caching is introduced, bring the if statement for it closer so it's less of a jump? I am sure they are both near jumps, but perhaps it can be optimized for jump prediction?